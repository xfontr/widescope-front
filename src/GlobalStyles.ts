import { createGlobalStyle } from "styled-components";

const GlobalStyles = createGlobalStyle`
  * {
    box-sizing: border-box;
  }

  body {
    margin: 0;
    min-height: 100vh;
  }

  ul {
    list-style: none;
    margin: 0;
    padding: 0;
  }

  button, input {
    font: inherit;
    border: none;
  }
`;

export default GlobalStyles;
