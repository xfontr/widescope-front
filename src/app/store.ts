import { configureStore, ThunkAction, Action } from "@reduxjs/toolkit";
import { userDataReducer } from "../store/slices/userData/userDataSlice";
import { userReducer } from "../store/slices/user/userSlice";

export const store = configureStore({
  reducer: {
    user: userReducer,
    userData: userDataReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
