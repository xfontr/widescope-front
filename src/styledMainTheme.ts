const styledMainTheme = {
  colors: {
    primary: "rgb(255, 255, 255)",
    primaryDarker: "rgb(238, 238, 238)",

    secondary: "rgb(25, 29, 35)",
    secondaryBrigther: "rgba(25, 29, 35, 0.7)",
  },

  shapes: {
    radiusSmall: "0.75rem",
    radiusBig: "1.5rem",

    shortShadow: "6px 6px 18px 6px rgba(0, 0, 0, 0.1)",
    longShadow: "10px 10px 30px 10px rgba(0, 0, 0, 0.1)",

    rectangularContainer: "0.8rem 1.5rem 0.8rem 1.5rem",
  },

  spacing: {
    gapSmall: "1rem",
    gapBig: "1.5rem",

    paddingSmall: "0.8rem",
    paddingBig: "1.5rem",
  },

  breakpoints: {
    small: "390px",
    big: "700pxpx",
    large: "1050px",
  },
};

export default styledMainTheme;
