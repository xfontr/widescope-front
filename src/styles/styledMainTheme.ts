const styledMainTheme = {
  colors: {
    primary: "rgb(255, 255, 255)",
    primaryDarker: "rgb(238, 238, 238)",

    secondary: "rgb(25, 29, 35)",
    secondaryBrigther: "rgba(25, 29, 35, 0.7)",

    error: "rgb(179, 120, 120)",
    success: "green",
  },

  shapes: {
    radiusSmall: "0.4rem",
    radiusBig: "1.5rem",

    shortShadow: "6px 6px 18px 6px rgba(0, 0, 0, 0.1)",
    longShadow: "10px 10px 30px 10px rgba(0, 0, 0, 0.1)",

    rectangularContainer: "0.7rem 1.6rem",
  },

  spacing: {
    gapSmall: "1.7rem",
    gapBig: "2.5rem",

    paddingSmall: "0.8rem",
    paddingBig: "1.8rem",
  },

  breakpoints: {
    small: "390px",
    big: "700px",
    large: "1250px",
  },
};

export default styledMainTheme;
