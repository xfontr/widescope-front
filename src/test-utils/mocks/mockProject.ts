import { IProject } from "../../types/project";

const mockProject: IProject = {
  id: "630e31ecb968115ba6a3e29f",
  name: "Project123",
  description: "The description",
  technologies: ["React", "Express"],
  repository: "www.google.com",
  author: "username",
  authorId: "id",
  logo: "ProjectLogo.png",
  logoBackup: "BackupProjectLogo.png",
  creationDate: "1662627402346",
};

export default mockProject;
