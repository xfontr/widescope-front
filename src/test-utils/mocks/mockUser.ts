import { IUser } from "../../types/user";

const mockUser: IUser = {
  id: "id",
  name: "name",
  email: "email@email.com",
  friends: ["friend_one", "friend_two"],
};

export default mockUser;
