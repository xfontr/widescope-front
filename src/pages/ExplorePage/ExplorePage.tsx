import { useEffect } from "react";
import { useAppSelector } from "../../app/hooks";
import { RootState } from "../../app/store";
import Projects from "../../components/Projects/Projects";
import useProjects from "../../hooks/useProjects/useProjects";

const ExplorePage = (): JSX.Element => {
  const { getAll } = useProjects();

  useEffect(() => {
    getAll();
  }, [getAll]);

  const allProjects = useAppSelector((state: RootState) => state.projects);

  return (
    <>
      <h2 className="page__title">
        These are the <span className="page__title--bold">latest projects</span>
      </h2>

      {allProjects.length && <Projects projects={allProjects} />}
      {!allProjects.length && <span>No projects found.</span>}
    </>
  );
};

export default ExplorePage;
