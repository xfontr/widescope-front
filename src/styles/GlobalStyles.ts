import { createGlobalStyle } from "styled-components";
import styledMainTheme from "./styledMainTheme";

const GlobalStyles = createGlobalStyle`
  * {
    box-sizing: border-box;
  }

  body {
    margin: 0;
    min-height: 100vh;
    background-color: ${styledMainTheme.colors.primaryDarker};
  }

  ul {
    list-style: none;
    margin: 0;
    padding: 0;
  }

  button, input {
    font: inherit;
    border: none;
  }
`;

export default GlobalStyles;
