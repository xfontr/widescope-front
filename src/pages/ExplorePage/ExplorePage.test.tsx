import { render as reactRender, waitFor } from "@testing-library/react";
import mockProject from "../../test-utils/mocks/mockProject";
import mockUser from "../../test-utils/mocks/mockUser";
import { render, screen } from "../../test-utils/render/customRender";
import { WrapperWithMockStore } from "../../test-utils/render/Wrapper";
import ExplorePage from "./ExplorePage";
import userEvent from "@testing-library/user-event";
import {
  GetAllProjects,
  UserProjects,
} from "../../hooks/types/useProjectTypes";
import { AxiosResponse } from "axios";
import { Provider } from "react-redux";
import { store } from "../../app/store";
import { ThemeProvider } from "styled-components";
import { BrowserRouter } from "react-router-dom";
import GlobalStyles from "../../styles/GlobalStyles";
import styledMainTheme from "../../styles/styledMainTheme";
import renderer from "react-test-renderer";

const mockAmountOfProjects = 9;
const mockListOfProjects = new Array(mockAmountOfProjects).fill(mockProject);
let mockResolvedValue: Partial<AxiosResponse<UserProjects | GetAllProjects>> = {
  data: {
    projects: {
      offset: 0,
      limit: 0,
      list: [
        ...mockListOfProjects,
        {
          ...mockProject,
          author: "Fake author",
          technologies: ["false", "false"],
        },
      ],
    },
  },
};

jest.mock("axios", () => ({
  get: () => mockResolvedValue,
}));

describe("Given a ExplorePage component", () => {
  describe("When instantiated", () => {
    test("Then it should match the snapshot", () => {
      const expectedPage = renderer.create(
        <Provider store={store}>
          <ThemeProvider theme={styledMainTheme}>
            <BrowserRouter>
              <GlobalStyles />
              <ExplorePage />
            </BrowserRouter>
          </ThemeProvider>
        </Provider>
      );

      expect(expectedPage).toMatchSnapshot();
    });

    test("Then it should show a title, the projects retreived and the pagination", async () => {
      render(<ExplorePage />);

      const heading = screen.getByRole("heading", {
        name: "These are the latest projects",
      });

      const cards = await screen.findAllByRole("heading", {
        name: mockProject.name,
        level: 3,
      });

      expect(heading).toBeInTheDocument();
      expect(cards).toHaveLength(mockAmountOfProjects + 1);

      const pagination = [
        screen.getByRole("button", { name: "»" }),
        screen.getByRole("button", { name: "«" }),
      ];

      pagination.forEach((page) => expect(page).toBeInTheDocument());
    });
  });

  describe("When intantiated but no projects are found", () => {
    test("Then it should show a message 'No projects found.'", async () => {
      reactRender(<ExplorePage />, { wrapper: WrapperWithMockStore });

      const cards = screen.queryByRole("heading", {
        name: mockProject.name,
        level: 3,
      });

      const message = await screen.findByText("No projects found.");

      expect(cards).toBeNull();
      expect(message).toBeInTheDocument();
    });
  });

  describe("When instantiated and the user clicks an author tag", () => {
    test("Then it should show only the projects of the selected user", async () => {
      mockResolvedValue = {
        data: {
          projectsByAuthor: {
            author: mockProject.author,
            total: 9,
            projects: [...mockListOfProjects],
          },
        },
      };

      render(<ExplorePage />);

      const author = await screen.findAllByText(mockUser.name);

      await userEvent.click(author[0]);

      const heading = await screen.findByRole("heading", {
        name: `Projects by ${mockUser.name}`,
      });

      const projectByAnotherAuthor = screen.queryByText("Fake author");
      const projectByThisAuthor = screen.getAllByText(mockProject.name);

      expect(heading).toBeInTheDocument();
      expect(projectByThisAuthor).toHaveLength(mockAmountOfProjects);

      await waitFor(() => {
        expect(projectByAnotherAuthor).not.toBeInTheDocument();
      });
    });

    describe("When instantiated and the user clicks a technology tag", () => {
      test("Then it should show only the projects of the selected technology", async () => {
        render(<ExplorePage />);

        const technologyTag = screen.getAllByText(mockProject.technologies[0]);
        await userEvent.click(technologyTag[0]);

        const currentQuery = screen.queryByText(
          `Searching by: Technology (${mockProject.technologies[0]})`
        );
        const projectTechnologies = screen.getAllByText(
          mockProject.technologies[0]
        );

        expect(currentQuery).toBeInTheDocument();
        expect(projectTechnologies).toHaveLength(mockAmountOfProjects);
      });
    });

    describe("When instantiated and the user sets a filter", () => {
      test("Then it should appear a link to reset the filters and see all projects again", async () => {
        render(<ExplorePage />);

        const author = await screen.findAllByText(mockUser.name);
        await userEvent.click(author[0]);
        const navigationLink = screen.getByText("« Keep exploring");

        expect(navigationLink).toBeInTheDocument();

        await userEvent.click(navigationLink);

        expect(navigationLink).not.toBeInTheDocument();

        const heading = await screen.findByRole("heading", {
          name: "These are the latest projects",
        });
        const projecstByThisAuthor = screen.getAllByText(mockProject.name);

        expect(heading).toBeInTheDocument();
        expect(projecstByThisAuthor).toHaveLength(mockAmountOfProjects + 1);
      });
    });
  });
});
