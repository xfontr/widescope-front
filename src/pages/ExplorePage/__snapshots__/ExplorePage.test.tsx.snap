// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a ExplorePage component When instantiated Then it should match the snapshot 1`] = `
Array [
  <h2
    className="page__title"
  >
    These are the
    <span
      className="page__title--bold"
    >
       latest projects
    </span>
  </h2>,
  <span>
    No projects found.
  </span>,
]
`;
